Improve Shoop python packaging (SHOOP-994)

suutari-ai opened the pull request:
See the commit messages for details.

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/737/
Test FAILed.

akx commented on build_resources.py:
```diff
@@ -0,0 +1,203 @@
+    if os.path.isdir(path):
+        tp = 'dir'
+        checksum.update(b'DIR:\n')
+        for item in os.listdir(path):
```
`sorted(os.listdir(path))` sounds like a good idea.

suutari-ai replied:
Ah, good catch. I had that at some point, but apparently it dropped at some refactoring phase. Thanks!

suutari-ai replied:
I pushed a fixup commit for that. Don't merge before I've rebased the fixup into the original commit.

suutari-ai commented:
The build failure seems to be
```
/tmp/hudson2367650515049819565.sh: 21: /tmp/hudson2367650515049819565.sh: .tox/py27/bin/flake8: not found
```
This happens, because tox no longer installs shoop[everything]. I'll look what I can do to this.

akx commented on doc/getting_started_dev.rst:
```diff
@@ -40,28 +40,44 @@ project set up to use an SQLite database. It is used by the test suite
+Building resources
+------------------
+
+Shoop uses Javascript and CSS resources that are compiled from various
```
```
Shoop uses JavaScript and CSS resources that are compiled using various
Node.js packages.  These resources are compiled automatically by ``setup.py``
when installing Shoop with pip, but if you make changes to the source files
(e.g. under ``shoop/admin/static_src``), the resources have to be rebuilt.

This can be done with
```

suutari-ai replied:
Thanks. Also this has a fixup commit now, waiting to be rebased...

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/738/
Test FAILed.

akx commented on setup.py:
```diff
@@ -4,15 +4,20 @@
 #
 # This source code is licensed under the AGPLv3 license found in the
 # LICENSE file in the root directory of this source tree.
+import distutils.command.build
```
All this new stuff in setup.py... Should some of it maybe be refactored into a separate module?

suutari-ai replied:
I thought about that, but importing stuff into setup.py is a bit problematic, since it's not in a package and implicit relative imports don't work in Python 3...

Maybe I'll figure out if it's easily doable...

akx replied:
`shoop_setup_utils.py`?

suutari-ai replied:
The problem was not that I couldn't figure out a name for that module. ;) (Even though naming is hard.)

suutari-ai replied:
Refactored all setup.py utils into shoop_setup_utils package.

akx commented on shoop/admin/generated_resources.txt:
```diff
@@ -0,0 +1,9 @@
+static/shoop_admin/js/picotable.js
```
Would be nice if these files had a header informing the user that this is an autogenerated manifest...

suutari-ai replied:
It is not autogenerated. It's a file that tells which resources will be generated by "npm run build" so that build_resources.py knows what to cache.

akx replied:
I have not the slightest doubt that someone will forget to update this, then.

suutari-ai replied:
Anyway, header would be nice still. The parser already ignores lines starting with a `#`, so I'll add a header.

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/739/
Test FAILed.

akx commented on MANIFEST.in:
```diff
@@ -1,2 +1,23 @@
+
+include tox.ini
+
+include build_resources.py nodejs_verify.py
```
Wait, uh, won't these be also copied into `site-packages` all willy-nilly?

suutari-ai replied:
Nope. Just to sdist. Not really sure why, though. Maybe because they're not inside any python package?

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/743/
Test FAILed.

akx commented:
So yeah, aside from the flake8 thing and lack of fixup-rebase, looks good, I think :)

suutari-ai commented:
Added following to Jenkins run script, so that flake8 gets installed:
```
.tox/${TOXENV}/bin/pip install shoop[coding-style]
```
Hopefully that will make this pass.

Also:
* rebased the fixups in
* moved more stuff shoop_setup_utils (mainly build_resources.py)
* fixed "npm run build" subprocess calls on Windows.

suutari-ai commented:
retest this please

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/744/
Test PASSed.

akx commented:
![pc749 cool beans](https://cloud.githubusercontent.com/assets/58669/8375262/5b989560-1c04-11e5-8c18-5cc9e9be9d4b.jpg)

Let's merge this. :D And great work obvs!
